src/fparser/two/symbol_table.py

```diff
@@ -0,0 +1,402 @@
+        is an existing stack then a new table is created and added to the
+        bottom.
+
+        :param str name: name of the scoping region.
+
+        '''
+        lname = name.lower()
+
+        if not self._scope_stack:
+            # We're not already inside a nested scope.
+            try:
+                table = self.lookup(lname)
+            except KeyError:
+                # Create a new, top-level symbol table with the supplied name.
+                table = self.add(lname)
+        else:
+            # We are already inside a scoping region so create a new table
+            # and setup its parent/child connections.
+            table = SymbolTable(lname, parent=self._scope_stack[-1])
+            self._scope_stack[-1].add_child(table)
+
+        # Finally, put the table on the stack of nested regions
+        self._scope_stack.append(table)
+
+    def exit_scope(self):
+        '''
+        Marks the end of the processing of the current scoping unit. Removes
+        the current symbol table from the stack.
+
+        :raises SymbolTableError: if there is no current scope from which to \
+                                  exit.
+        '''
+        if not self._scope_stack:
+            raise SymbolTableError("exit_scope() called but no current scope "
+                                   "exists.")
+        self._scope_stack.pop(-1)
+
+
+class SymbolTable(object):
+    '''
+    Class implementing a single symbol table.
+
+    :param str name: the name of this scope. Will be the name of the \
+                     associated module or routine.
+    :param parent: the symbol table within which this one is nested (if any).
+    :type parent: :py:class:`fparser.two.symbol_table.SymbolTable.Symbol`
+
+    '''
+    # TODO #201 add support for other symbol properties (kind, shape
+    # and visibility). We may need a distinct Symbol class so as to provide
+    # type checking for the various properties.
+    Symbol = namedtuple("Symbol", "name primitive_type")
+
+    def __init__(self, name, parent=None):
+        self._name = name.lower()
+        # Symbols defined in this scope.
+        self._symbols = {}
+        # Modules imported into this scope.
+        self._modules = {}
+        # Reference to a SymbolTable that contains this one (if any). Actual
+        # value (if any) is set via setter method.
+        self._parent = None
+        self.parent = parent
+        # Symbol tables nested within this one.
+        self._children = []
+
+    def __str__(self):
+        header = "===========\n"
+        symbols = "Symbols:\n"
+        if self._symbols:
+            symbols += "\n".join(list(self._symbols.keys())) + "\n"
+        uses = "Used modules:\n"
+        if self._modules:
+            uses += "\n".join(list(self._modules.keys())) + "\n"
+        return ("{0}Symbol Table '{1}'\n".format(header, self._name) +
+                symbols + uses + header)
+
+    def add_symbol(self, name, primitive_type):
```

Comment:
As mentioned in the previous comment, this method only adds one type of symbol. I think it should have a name reflecting the type of symbol being added e.g. `add_variable_symbol` (or the implementation changed so that all symbols of different types are added here).

Reply:
I was going to extend this in future. I'll add a TODO.

Reply:
Actually, looking at the code I think you're right and I've renamed it to `add_data_symbol`.